File: 0003_LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ret = 0
        left, right = 0,0
        mem = set()
        while left < len(s) and right < len(s):
            if s[right] not in mem:
                mem.add(s[right])
                right+=1
            else: 
                mem.remove(s[left])
                left+=1
            ret = max(ret, len(mem))
        return ret
    

File: 0003_LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.js
function lengthOfLongestSubstring (str) {
	let mem = new Set();
	let ret = 0;
	let [left, right] = [0,0]
	while ( left < str.length && right < str.length) {
		if (!mem.has(str[right])) {
			mem.add(str[right]);
			right++
		} else {
			mem.delete(str[left])
			left++
		}
		ret = Math.max(ret, mem.size)
	}
	return ret
}